src/Providers/Models/DTO/ModelRequirements.php

```diff
@@ -88,6 +91,192 @@ public function getRequiredOptions(): array
+                    $hasFunctionMessageParts = true;
+                }
+            }
+        }
+
+        //
+        // Convert ModelConfig to RequiredOptions (moved from ModelConfig::toRequiredOptions)
+        $requiredOptions = [];
+
+        // Map properties that have corresponding OptionEnum values
+        if ($modelConfig->getOutputModalities() !== null) {
+            $requiredOptions[] = new RequiredOption(
+                OptionEnum::outputModalities(),
+                $modelConfig->getOutputModalities()
+            );
+        }
+
+        if ($modelConfig->getSystemInstruction() !== null) {
+            $requiredOptions[] = new RequiredOption(
+                OptionEnum::systemInstruction(),
+                $modelConfig->getSystemInstruction()
+            );
+        }
+
+        if ($modelConfig->getCandidateCount() !== null) {
+            $requiredOptions[] = new RequiredOption(
+                OptionEnum::candidateCount(),
+                $modelConfig->getCandidateCount()
+            );
+        }
+
+        if ($modelConfig->getMaxTokens() !== null) {
+            $requiredOptions[] = new RequiredOption(
+                OptionEnum::maxTokens(),
+                $modelConfig->getMaxTokens()
+            );
+        }
+
+        if ($modelConfig->getTemperature() !== null) {
+            $requiredOptions[] = new RequiredOption(
+                OptionEnum::temperature(),
+                $modelConfig->getTemperature()
+            );
+        }
+
+        if ($modelConfig->getTopP() !== null) {
+            $requiredOptions[] = new RequiredOption(
+                OptionEnum::topP(),
+                $modelConfig->getTopP()
+            );
+        }
+
+        if ($modelConfig->getTopK() !== null) {
+            $requiredOptions[] = new RequiredOption(
+                OptionEnum::topK(),
+                $modelConfig->getTopK()
+            );
+        }
+
+        if ($modelConfig->getOutputMimeType() !== null) {
+            $requiredOptions[] = new RequiredOption(
+                OptionEnum::outputMimeType(),
+                $modelConfig->getOutputMimeType()
+            );
+        }
+
+        if ($modelConfig->getOutputSchema() !== null) {
+            $requiredOptions[] = new RequiredOption(
+                OptionEnum::outputSchema(),
+                $modelConfig->getOutputSchema()
+            );
+        }
+
+        // Step 5: Add additional options based on message analysis
+        if ($hasFunctionMessageParts) {
+            $requiredOptions[] = new RequiredOption(OptionEnum::functionDeclarations(), true);
+        }
+
+        // Add input modalities if we have any inputs
+        if (!empty($inputModalities)) {
+            // Remove duplicates
+            $inputModalities = array_unique($inputModalities, SORT_REGULAR);
+            $requiredOptions[] = new RequiredOption(OptionEnum::inputModalities(), array_values($inputModalities));
+        }
```

Comment:
This is missing the logic to ensure the `RequiredOption` is only added once to the array.

I think this is best solved by reinstating the private `includeInRequiredOptions` method from before, but now in this class.